### PR TITLE
Feat/Interactive Schedule

### PR DESCRIPTION
**Pull request by Anon-136**
## Why did you create this PR
- It's beautiful
Screen Shot 2564-11-15 at 00 27 23

## What did you do
Screen Shot 2564-12-14 at 12 52 02

- Implement interactive schedule
Screen Shot 2564-12-14 at 12 54 21

- Add color picker
Screen Shot 2564-12-14 at 13 01 55

- Change class description from classroom to section
Screen Shot 2564-11-15 at 17 25 04 -> Screen Shot 2564-11-15 at 17 21 34

- Add type of class
Screen Shot 2564-12-14 at 12 57 24

- Fix section card show `GenEdChip` for specific section
Please check GEN PSY [dev](https://dev.cugetreg.com/S/courses/3800101) vs [beta](https://beta.cugetreg.com/S/courses/3800101)
- Also add `GenEdChip` for selected section for selected course card
Screen Shot 2564-12-14 at 14 19 49

- Adjust two lines table to top vertical align
- Update types and CourseCartStore
- The color will be randomly picked for each course when you initially add to the cart
- Automatically add color when sync from store
- Also improve responsive on schedule page

- Change slide threshold on selected course card 

## Demo
[https://dev.cugetreg.com](https://dev.cugetreg.com)

## Checklist
- [x] Deploy a demo
- [x] Check browsers compatibility
- [ ] Wrote coverage tests

 
P.S. I'll write test later

### CODE REVIEW

**Comment by new5558** [comment]
Why do we have "เลือกสี" button since users cannot see a preview of the selected course anyway?
Should we immediately close the color picker when the user selected the color?

image

File: src/modules/Schedule/components/ColorPicker/utils/getNewColor.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Course } from '@thinc-org/chula-courses'
+import { difference, random } from 'lodash'
+
+import { CourseCartItem } from '@/store'
+
+import { OTHER_COLORS } from '../constants'
+
+export function getNewColor(currentItems: CourseCartItem[], newCourse: Course) {
```
Comment (new5558) [inline src/modules/Schedule/components/ColorPicker/utils/getNewColor.ts:8]: Can we JSdoc the reason we need this function (for when user add course first time and for when user sync with previously uploaded gDrive )?

File: src/modules/Schedule/components/Schedule/components/ClassCard.tsx [hunk 2]
```diff
@@ -56,7 +70,7 @@ export function ClassCard({ scheduleClass }: ClassCardProps) {
         {isWide && courseNo} {abbrName}
       </ClassCardTypography>
       <ClassCardTypography variant="caption">
-        {teachers.join(', ')} | {building} {room}
+        Sec {sectionNo} - {teachers.join(', ')}
```
Comment (new5558) [inline src/modules/Schedule/components/Schedule/components/ClassCard.tsx:73]: Should document this change on this Github PR detail?

File: src/modules/Schedule/components/ScheduleTable/components/ScheduleTableCard/index.tsx [hunk 3]
```diff
@@ -66,39 +68,41 @@ export const ScheduleTableCard = observer(({ item, index, overlaps }: ScheduleTa
...
     }
   }, [match])
-  const onDragEnd = (e: MouseEvent, { offset, point }: PanInfo) => {
+
+  function onDragEnd(e: MouseEvent, { offset, point }: PanInfo) {
     // cancelling drag event due to scrolling
     if (point.x == 0 && point.y == 0) {
       return
     }
+    const DRAG_THRESHHOLD = 25
```
Comment (new5558) [inline src/modules/Schedule/components/ScheduleTable/components/ScheduleTableCard/index.tsx:77]: Should we also document this change on this Github PR detail?

File: src/modules/Schedule/components/ColorPicker/constants.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+export const GENED_COLORS = ['pink', 'secondary', 'purple', 'green'] as const
+export const OTHER_COLORS = ['orange', 'teal', 'blue', 'indigo', 'primary'] as const
+export const SCHEDULE_COLORS = [...GENED_COLORS, ...OTHER_COLORS] as const
+
+export type GenEdColor = typeof GENED_COLORS[number]
```
Comment (new5558) [inline src/modules/Schedule/components/ColorPicker/constants.ts:5]: Where we use this?

**Review by new5558 — COMMENTED** [review]
Looks beautiful!

File: src/modules/Schedule/components/Schedule/components/ClassCard.tsx (hunk 2 above)
Comment (Anon-136) [reply src/modules/Schedule/components/Schedule/components/ClassCard.tsx:73]: oh yeah, thanks for mentions

File: src/modules/Schedule/components/ColorPicker/constants.ts (hunk 4 above)
Comment (Anon-136) [reply src/modules/Schedule/components/ColorPicker/constants.ts:5]: I don't remember anymore lol
So let me remove that

File: src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null)
+  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
+    setAnchorEl(event.currentTarget)
+  }
+  const open = !!anchorEl
+  const handleClose = () => {
+    setAnchorEl(null)
+  }
+  const setColor = (scheduleColor: ScheduleColor) => {
+    courseCartStore.changeColor(classKey, scheduleColor)
```
Comment (new5558) [inline src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:21]: Can we also measure here using useLog to know which color is user preferred most?

**Comment by natTP** [comment]
> Why do we have "เลือกสี" button since users cannot see a preview of the selected course anyway? Should we immediately close the color picker when the user selected the color?
> 
> image

There _is_ a preview, the color changes immediately. But I understand why users wouldn’t see that on mobile.
I think the user should be able to choose when to close the dialog by themselves, because even without a preview, they might want to think about what to choose and try selecting different colors.

File: src/modules/Schedule/components/ScheduleTable/components/ScheduleTableCard/index.tsx [hunk 6]
```diff
@@ -140,27 +144,44 @@ function CardHeader({ item }: CardComponentProps) {
...
-      </Typography>
-      <Hidden smDown>
-        <SectionSelect item={item} />
-        <Spacer />
-      </Hidden>
+    <Stack direction="row" pt={1} my={0.5} pr={2}>
+      <Stack direction="row" flex={1} justifyContent="space-between" alignItems="flex-start">
+        <Grid container columnGap={2} alignItems="center" py={0.5}>
+          <Grid item>
+            <StyledLink href={buildLink(`/courses/${item.courseNo}`)}>
```
[the remark this answers is not in the document]
Comment (Anon-136) [reply src/modules/Schedule/components/ScheduleTable/components/ScheduleTableCard/index.tsx]: done krub

File: src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts (hunk 5 above)
Comment (Anon-136) [reply src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:25]: Done.
Please check if i use it properly

File: src/modules/Schedule/components/ColorPicker/utils/getNewColor.ts (hunk 1 above)
Comment (Anon-136) [reply src/modules/Schedule/components/ColorPicker/utils/getNewColor.ts:19]: Done
Should be more clear now

**Comment by natTP** [comment]
ว่าแต่ เพิ่งเห็นว่าลำดับการเรียงสีใน dialog มันไม่ใช่ rainbow 555555555

File: src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts (hunk 5 above)
Comment (new5558) [reply src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:25]: Will do it tommorow!

**Comment by saenyakorn** [comment]
scroll ไม่ได้เวลา dialog มันเปิดขึ้นมา

File: src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import { useLog } from '@/common/context/Analytics/hooks/useLog'
+import { ScheduleColor } from '@/modules/Schedule/components/ColorPicker/constants'
+import { CourseCartItem, courseCartStore } from '@/store'
+
+export type ColorClassKey = Pick<
+  CourseCartItem,
+  'courseNo' | 'studyProgram' | 'academicYear' | 'semester' | 'abbrName' | 'color'
+>
+
+export function useColorPicker(classKey: ColorClassKey) {
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:13]: please write test for this hook

File: src/modules/Schedule/components/ColorPicker/index.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import { Box, Button, ButtonProps, Popover, Stack, Typography, useTheme } from '@mui/material'
+import { useTranslation } from 'react-i18next'
+
+import { ColorClassKey } from '@/modules/Schedule/components/ColorPicker/hooks/useColorPicker'
+
+import { ScheduleColor, SCHEDULE_COLORS } from './constants'
+import { useColor } from './hooks/useColor'
+
+interface ColorPickerProps {
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/index.tsx:9]: please move to `src/modules/Schedule/components/ColorPicker/types.ts`

File: src/modules/Schedule/components/ColorPicker/index.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  open: boolean
+  handleClose: () => void
+  scheduleClass: ColorClassKey
+  anchorEl: HTMLElement | null
+  setColor: (color: ScheduleColor) => void
+  selectedColor: ScheduleColor | undefined
+}
+
+export const ColorPicker = (props: ColorPickerProps) => {
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/index.tsx:18]: please write a test for this component

File: src/modules/Schedule/components/ColorPicker/index.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        ':hover': {
+          background: theme.palette.primaryRange[30],
+        },
+      }}
+      {...buttonProps}
+    >
+      <Circle color={scheduleColor} />
+    </Button>
+  )
+}
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/index.tsx:77]: please move to `src/modules/Schedule/components/ColorButton`

File: src/modules/Schedule/components/ColorPicker/index.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        height: size,
+        minWidth: size,
+        minHeight: size,
+        background: scheme.background,
+        border: `1px solid ${scheme.border}`,
+        borderRadius: 100,
+      }}
+    />
+  )
+}
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/index.tsx:99]: please move this to `src/modules/Schedule/components/ColorCircle` and rename this component to `CircleColor`. (`Circle` is too general)

Comment (saenyakorn) [reply src/modules/Schedule/components/ColorPicker/index.tsx:99]: Also please write a simple test for this component

File: src/modules/Schedule/components/ColorPicker/index.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+      sx={{
+        width: size,
+        height: size,
+        minWidth: size,
+        minHeight: size,
+        background: scheme.background,
+        border: `1px solid ${scheme.border}`,
+        borderRadius: 100,
+      }}
+    />
```
Comment (saenyakorn) [inline src/modules/Schedule/components/ColorPicker/index.tsx:97]: to render `Box` component is too expensive, especially for looped components. Can you please use `emotion` or something else instead?

**Review by saenyakorn — CHANGES_REQUESTED** [review]

File: src/modules/Schedule/components/ColorPicker/index.tsx (hunk 11 above)
Comment (paphonb) [reply src/modules/Schedule/components/ColorPicker/index.tsx:99]: Personally I think the name is fine

Comment (paphonb) [reply src/modules/Schedule/components/ColorPicker/index.tsx:99]: Should we really test this? It's literally just styles.

File: src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts (hunk 7 above)
Comment (new5558) [reply src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:13]: We need testing tutorial

Comment (Anon-136) [reply src/modules/Schedule/components/ColorPicker/hooks/useColorPicker.ts:12]: Yes we do

File: src/modules/Schedule/components/ColorPicker/index.tsx (hunk 12 above)
Comment (Anon-136) [reply src/modules/Schedule/components/ColorPicker/index.tsx]: Moved to `styled` instead

**Comment by Anon-136** [comment]
> scroll ไม่ได้เวลา dialog มันเปิดขึ้นมา

I can enable the scroll but it will no longer be attached to the button
I'll merge to dev so you can see

**Comment by Anon-136** [comment]
> > scroll ไม่ได้เวลา dialog มันเปิดขึ้นมา
> 
> I can enable the scroll but it will no longer be attached to the button I'll merge to dev so you can see

Fixed, used `Popper` instead

File: src/store/courseCart.ts [hunk 13]
```diff
@@ -257,9 +268,14 @@ export class CourseCart implements CourseCartProps {
...
-    const newItem: CourseCartItem = { ...course, selectedSectionNo, isHidden: false }
+    const newItem: CourseCartItem = {
+      ...course,
+      selectedSectionNo,
+      isHidden: false,
+      color: getNewColor(this.shopItems, course),
+    }
     const foundIndex = this.shopItems.findIndex((item) => isSameKey(item, newItem))
-    if (foundIndex != -1) this.shopItems[foundIndex] = newItem
+    if (foundIndex != -1) this.shopItems[foundIndex] = { ...this.shopItems[foundIndex], selectedSectionNo }
```
Comment (saenyakorn) [inline src/store/courseCart.ts:278]: Did you mean?
```suggestion
    if (foundIndex !== -1) this.shopItems[foundIndex] = { ...this.shopItems[foundIndex], selectedSectionNo }
```

**Review by saenyakorn — APPROVED** [review]

File: src/store/courseCart.ts (hunk 13 above)
Comment (Anon-136) [reply src/store/courseCart.ts]: oh yeah

**Pull request merged**